Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,239 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftServiceLifecycle open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftServiceLifecycle project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftServiceLifecycle project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import UnixSignals
+
+/// A ``ServiceRunner`` is responsible for running a number of services, setting up signal handling and shutting down services in the correct order.
+public actor ServiceRunner: Sendable {
+    /// The internal state of the ``ServiceRunner``.
+    private enum State {
+        /// The initial state of the runner.
+        case initial
+        /// The state once ``ServiceRunner/run()`` has been called.
+        case running
+        /// The state once ``ServiceRunner/run()`` has finished.
+        case finished
+    }
+
+    /// The services to run.
+    private let services: [any Service]
+    /// The runner's configuration.
+    private let configuration: ServiceRunnerConfiguration
+    /// The logger.
+    private let logger: Logger
+
+    /// The current state of the runner.
+    private var state: State = .initial
+
+    /// Initializes a new ``ServiceRunner``.
+    ///
+    /// - Parameters:
+    ///   - services: The services to run.
+    ///   - configuration: The runner's configuration.
+    ///   - logger: The logger.
+    public init(
+        services: [any Service],
+        configuration: ServiceRunnerConfiguration,
+        logger: Logger
+    ) {
+        self.services = services
+        self.configuration = configuration
+        self.logger = logger
+    }
+
+    /// Runs all the services by spinning up a child task per service.
+    /// Furthermore, this method sets up the correct signal handlers
+    /// for graceful shutdown.
+    public func run() async throws {
+        switch self.state {
+        case .initial:
+            self.state = .running
+            try await self._run()
+
+            switch self.state {
+            case .initial, .finished:
+                fatalError("ServiceRunner is in an invalid state \(self.state)")
+
+            case .running:
+                self.state = .finished
+            }
+
+        case .running:
+            throw ServiceRunnerError.alreadyRunning()
+
+        case .finished:
+            throw ServiceRunnerError.alreadyFinished()
```

Comment:
Adding the file and line number here is nice for us as library authors, but for users of the lib it would be _much_ nicer to get the file and line number where the adopter has invoked the run method. Because in those cases adopters have messed up their run methods and it would be extremely cool to tell them directly where.

Reply:
I guess we could just forward `line` and `file` here.